lint/rule.go

```diff
@@ -14,10 +14,304 @@
+
+// ServiceRule defines a lint rule that is run on each service.
+type ServiceRule struct {
+	Name RuleName
+	URI  string
+
+	// LintService accepts a ServiceDescriptor and lints it.
+	LintService func(*desc.ServiceDescriptor) []Problem
+
+	noPositional struct{}
+}
+
+// GetName returns the name of the rule.
+func (r *ServiceRule) GetName() RuleName {
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *ServiceRule) GetURI() string {
+	return r.URI
+}
+
+// Lint accepts a FileDescriptor and lints every service in the file.
+func (r *ServiceRule) Lint(fd *desc.FileDescriptor) []Problem {
+	problems := []Problem{}
+	for _, service := range fd.GetServices() {
+		problems = append(problems, r.LintService(service)...)
+	}
+	return problems
+}
+
+// MethodRule defines a lint rule that is run on each method.
+type MethodRule struct {
+	Name RuleName
+	URI  string
+
+	// LintMethod accepts a MethodDescriptor and lints it.
+	LintMethod func(*desc.MethodDescriptor) []Problem
+
+	noPositional struct{}
+}
+
+// GetName returns the name of the rule.
+func (r *MethodRule) GetName() RuleName {
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *MethodRule) GetURI() string {
+	return r.URI
+}
+
+// Lint accepts a FileDescriptor and lints every method in the file.
+func (r *MethodRule) Lint(fd *desc.FileDescriptor) []Problem {
+	problems := []Problem{}
+	for _, service := range fd.GetServices() {
+		for _, method := range service.GetMethods() {
+			problems = append(problems, r.LintMethod(method)...)
+		}
+	}
+	return problems
+}
+
+// EnumRule defines a lint rule that is run on each enum.
+type EnumRule struct {
+	Name RuleName
+	URI  string
+
+	// LintEnum accepts a EnumDescriptor and lints it.
+	LintEnum func(*desc.EnumDescriptor) []Problem
+
+	noPositional struct{}
+}
+
+// GetName returns the name of the rule.
+func (r *EnumRule) GetName() RuleName {
+	return r.Name
+}
+
+// GetURI returns the URI where the applicable guideline is documented.
+func (r *EnumRule) GetURI() string {
+	return r.URI
+}
+
+// Lint accepts a FileDescriptor and lints every enum in the file
+// (including enums nested within messages).
+func (r *EnumRule) Lint(fd *desc.FileDescriptor) []Problem {
+	problems := []Problem{}
+
+	// Lint enums that are at the top level of the file.
+	for _, enum := range fd.GetEnumTypes() {
+		problems = append(problems, r.LintEnum(enum)...)
+	}
+
+	// Lint enums that are nested within messages.
+	for _, message := range getAllMessages(fd) {
+		for _, enum := range message.GetNestedEnumTypes() {
+			problems = append(problems, r.LintEnum(enum)...)
+		}
+	}
+	return problems
+}
+
+// ruleIsEnabled returns true if the rule is enabled (not disabled by the comments
+// for the given descriptor or its file), false otherwise.
+func ruleIsEnabled(rule ProtoRule, d desc.Descriptor) bool {
+	directive := fmt.Sprintf("api-linter: %s=disabled", rule.GetName())
```

Comment:
If a user wants to disable two rules, he/she needs to write
```
api-linter: a=disabled
api-linter: b=disabled
```
but not
```
api-linter: a,b=disable
```
right?

Reply:
In users' perspective, I prefer the second way. But we can discuss more and if needed, make the changes in another PR. Thanks!

Reply:
This mirrors the previous behavior, and can be done later.